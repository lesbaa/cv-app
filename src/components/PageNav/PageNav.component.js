/* global window */

import React from 'react'
import PropTypes from 'prop-types'
import Link from 'next/link'
import MoreVerticalIcon from 'react-feather/dist/icons/more-vertical'
import styles from './PageNav.styles'

const PageNav = ({
  router,
  slides,
}) => (
  <div className="PageNav">
    <span className="chevron">
      <MoreVerticalIcon
        strokeWidth="0.5px"
        size="3em"
      />
    </span>
    <ul className="nav-links">
      {mapSlides(slides, router)}
    </ul>
    <style jsx>{styles}</style>
  </div>
)

PageNav.propTypes = {
  router: PropTypes.any,
  slides: PropTypes.array,
}

PageNav.defaultProps = {
  router: {},
  slides: {
    hello: {
      altText: 'Oops, something seems to have gone awry in here!',
    },
  },
}

function mapSlides(slides, router) {
  return Object.entries(slides)
    .sort((
      [, { order: a }],
      [, { order: b }],
    ) => a - b)
    .map(([slidename, { altText }]) => {
      const isActiveLink = router.query.slidename === slidename
      const navItemClassnames = [
        'nav-item',
        isActiveLink && 'active',
      ]
        .filter(Boolean)
        .join(' ')

      const Element = isActiveLink
        ? props => <span {...props} />
        : Link

      return (
        <li className={navItemClassnames}>
          <Element
            href={`/CVSlide?slidename=${slidename}`}
            as={`/cv/${slidename}`}
          >
            <a title={altText}>
              <img
                src={`/static/img/nav-icons/${slidename}.svg`}
                alt={`An icon for the ${slidename} page, ${altText}`}
              />
            </a>
          </Element>
          <style jsx>{styles}</style>
        </li>
      )
    })
}

export default PageNav
