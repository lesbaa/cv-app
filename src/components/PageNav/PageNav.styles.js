import css from 'styled-jsx/css'

export default css`
  .PageNav {
    position: absolute;
    right: -4em;
    top: 0;
    bottom: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    transform: translateX(0);
    mix-blend-mode: color-burn;
    z-index: 2;
  }

  .nav-links {
    display: inline-block;
    margin: auto 0;
    padding: 0;
  }

  .chevron {
    vertical-align: middle;
    line-height: 100%;
    padding: 0.5em 0;
    border: 1px solid #444;
    border-right-width: 0;
    border-radius: 100% 0 0 100%;
  }

  .PageNav,
  .chevron,
  .nav-item,
  .nav-item a {
    transition: 0.25s opacity, 0.5s transform;
  }

  .nav-item {
    margin: 0.5em;
    display: block;
  }

  .nav-item:hover {
    opacity: 0.6;
  }

  .nav-item.active {
    transform: scale(1.4);
    opacity: 0.6;
  }

  .nav-item.active a {
    cursor: default;
  }

  .nav-item > :global(svg) {
    height: 2.5em;
    width: 2.5em;
  }

  .PageNav:hover {
    transform: translateX(-70%);
  }


  .PageNav:hover .chevron {
    opacity: 0;
  }
`
