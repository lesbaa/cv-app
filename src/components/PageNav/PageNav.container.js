import { withRouter } from 'next/router'
import { connect } from 'react-redux'
import { compose } from 'ramda'
import PageNav from './PageNav.component'

const mapStateToProps = ({
  slides,
}) => ({
  slides,
})

const enhance = compose(
  withRouter,
  connect(mapStateToProps)
)

export default enhance(PageNav)
