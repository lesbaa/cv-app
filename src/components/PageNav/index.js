import PageNav from './PageNav.container'

export default PageNav
