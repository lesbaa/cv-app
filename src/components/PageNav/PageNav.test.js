/* eslint-env jest */
/* global window */
import React from 'react'
import { shallow } from 'enzyme'
import PageNav from './PageNav.component'

const mockSlides = {
  'les is so cool': {
    order: 1,
    altText: 'ALT_TEXT',
  },
  'really?': {
    order: 3,
    altText: 'ALT_TEXT',
  },
  'yeah, he\'s like the coolest cat out there, man': {
    order: 7,
    altText: 'ALT_TEXT',
  },
  'well, we should probably give him a job': {
    order: 5,
    altText: 'ALT_TEXT',
  },
  'fo sho?': {
    order: 4,
    altText: 'ALT_TEXT',
  },
  'FO. SHO.': {
    order: 2,
    altText: 'ALT_TEXT',
  },
  'you\'re god damn hired, Les.': {
    order: 6,
    altText: 'ALT_TEXT',
  },
}

describe('PageNav', () => {
  const wrapperOne = shallow(
    <PageNav
      slides={mockSlides}
      router={{
        query: {
          slidename: 'les is so cool',
        },
      }}
    />
  )

  const wrapperTwo = shallow(
    <PageNav
      slides={mockSlides}
      router={{
        query: {
          slidename: 'fo sho?',
        },
      }}
    />
  )

  it('matches its snapshots', () => {
    expect(wrapperOne).toMatchSnapshot('wrapperOne')
    expect(wrapperTwo).toMatchSnapshot('wrapperTwo')
  })

  it('renders the links in the correct order', () => {
    wrapperOne
      .find('.nav-item')
      .forEach((node, index) => {
        const expectedKey = String(index + 1)
        expect(expectedKey).toBe(node.key())
      })
  })

  it('renders the links with the correct props', () => {
    const {
      as,
      href,
    } = wrapperOne
      .find('Link')
      .at(3) // chosen at random...
      .props()

    expect(href).toBe('/CVSlide?slidename=well, we should probably give him a job')
    expect(as).toBe('/cv/well, we should probably give him a job')
  })

  it('renders a span element instead of a Link for the active link', () => {
    // this is not ideal and makes the test fragile, however, see here
    // https://stackoverflow.com/questions/52221341/enzyme-findselector-does-not-seem-to-find-a-selector
    // https://github.com/airbnb/enzyme/issues/1807
    const isSpan = wrapperOne
      .find('.active-nav-option')
      .html()
      .match('<span')

    expect(Boolean(isSpan)).toBe(true)
  })

})
