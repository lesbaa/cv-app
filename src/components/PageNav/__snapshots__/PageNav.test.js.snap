// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageNav matches its snapshots: wrapperOne 1`] = `ShallowWrapper {}`;

exports[`PageNav matches its snapshots: wrapperTwo 1`] = `ShallowWrapper {}`;
